File: examples/php/Case04Test.php
<?php

require_once __DIR__ . '/Case04.php';

function case04Require(bool $condition, string $message): void
{
    if (!$condition) throw new RuntimeException($message);
}

$weak = new Case04RewardStore();
$weakAck = new Case04DeliveryAck();
case04WeakEffectThenAck($weak, $weakAck, 'E', 'User42', 100, true);
case04Require(!isset($weakAck->accepted['E']), 'pre-ACK crash was acknowledged');
case04WeakEffectThenAck($weak, $weakAck, 'E', 'User42', 100, false);
case04Require($weak->state['balance']['User42'] === 200, 'weak D1+D2 did not repeat');

$ackFirst = new Case04RewardStore();
$firstAck = new Case04DeliveryAck();
case04AckBeforeEffect($ackFirst, $firstAck, 'E', 'User42', 100, true);
case04Require(isset($firstAck->accepted['E']) && !isset($ackFirst->state['balance']['User42']),
    'ACK-first did not expose lost effect');

$strong = new Case04RewardStore();
case04Require(case04ApplyEventOnce($strong, 'E', 'User42', 100), 'new E suppressed');
case04Require(!case04ApplyEventOnce($strong, 'E', 'User42', 100), 'D2(E) applied');
$strongAck = new Case04DeliveryAck();
case04HandleRewardDelivery($strong, $strongAck, 'E', 'User42', 100);
case04Require($strong->state['balance']['User42'] === 100 && isset($strongAck->accepted['E']),
    'redelivery did not finish safely');

$split = new Case04RewardStore();
$marker = [];
case04SplitRewardAndMarker($split, $marker, 'E', 'User42', 100, true);
case04SplitRewardAndMarker($split, $marker, 'E', 'User42', 100, false);
case04Require($split->state['balance']['User42'] === 200, 'split marker did not repeat effect');

File: examples/c/case04_test.c
#include <assert.h>
#include <threads.h>
#include "case04.c"

typedef struct { Case04RewardStore *store; int result; } Case04ConcurrentDelivery;

static int case04_concurrent_apply(void *arg) {
  Case04ConcurrentDelivery *delivery = arg;
  delivery->result = case04_apply_event_once(delivery->store, "E-concurrent", 100);
  return 0;
}

int main(void) {
    Case04RewardStore weak = {0};
    Case04DeliveryAck weak_ack = {0};
    case04_weak_effect_then_ack(&weak, &weak_ack, 100, 1);
    assert(!weak_ack.accepted);
    case04_weak_effect_then_ack(&weak, &weak_ack, 100, 0);
    assert(weak.user42_balance == 200);

    Case04RewardStore ack_first = {0};
    Case04DeliveryAck first_ack = {0};
    case04_ack_before_effect(&ack_first, &first_ack, 100, 1);
    assert(first_ack.accepted && ack_first.user42_balance == 0);

    Case04RewardStore strong = {0};
    assert(case04_apply_event_once(&strong, "E", 100) == 1);
    assert(case04_apply_event_once(&strong, "E", 100) == 0);
    assert(strong.user42_balance == 100);
    Case04DeliveryAck strong_ack = {0};
    assert(case04_handle_reward_delivery(&strong, &strong_ack, "E", 100) == 0);
  assert(strong_ack.accepted && strong.user42_balance == 100);

  Case04RewardStore concurrent = {0};
  Case04ConcurrentDelivery deliveries[8];
  thrd_t threads[8];
  for (size_t i = 0; i < 8; i++) {
    deliveries[i] = (Case04ConcurrentDelivery){.store = &concurrent};
    assert(thrd_create(&threads[i], case04_concurrent_apply, &deliveries[i]) == thrd_success);
  }
  int applied = 0;
  for (size_t i = 0; i < 8; i++) {
    assert(thrd_join(threads[i], NULL) == thrd_success);
    applied += deliveries[i].result;
  }
  assert(applied == 1 && concurrent.user42_balance == 100);

    Case04RewardStore split = {0};
    Case04Marker marker = {0};
    case04_split_reward_and_marker(&split, &marker, "E", 100, 1);
    case04_split_reward_and_marker(&split, &marker, "E", 100, 0);
    assert(split.user42_balance == 200);
    return 0;
}
